File: app/src/main/java/tanzent/cassette/misc/menu/SongPopupListener.java
package tanzent.cassette.misc.menu;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.PopupMenu;
import android.view.MenuItem;

import com.soundcloud.android.crop.Crop;

import java.util.ArrayList;
import java.util.Collections;

import tanzent.cassette.App;
import tanzent.cassette.R;
import tanzent.cassette.bean.misc.CustomThumb;
import tanzent.cassette.bean.mp3.Song;
import tanzent.cassette.service.Command;
import tanzent.cassette.service.MusicService;
import tanzent.cassette.theme.Theme;
import tanzent.cassette.ui.Tag;
import tanzent.cassette.ui.dialog.AddtoPlayListDialog;
import tanzent.cassette.util.Constants;
import tanzent.cassette.util.MediaStoreUtil;
import tanzent.cassette.util.PlayListUtil;
import tanzent.cassette.util.SPUtil;
import tanzent.cassette.util.ToastUtil;
import tanzent.cassette.util.Util;

import static com.afollestad.materialdialogs.DialogAction.POSITIVE;

/**
 * Created by Remix on 2018/3/5.
 */

public class SongPopupListener
        implements PopupMenu.OnMenuItemClickListener {
    private String mPlayListName;
    private boolean mIsDeletePlayList;
    private Song mSong;
    private Activity mActivity;
    private Tag mTag;

    public SongPopupListener(Activity activity, Song song, boolean isDeletePlayList, String playListName) {
        mIsDeletePlayList = isDeletePlayList;
        mPlayListName = playListName;
        mSong = song;
        mActivity = activity;
        mTag = new Tag(activity, mSong);
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_next:
                Intent intent = new Intent(MusicService.ACTION_CMD);
                intent.putExtra("Control", Command.ADD_TO_NEXT_SONG);
                intent.putExtra("song", mSong);
                Util.sendLocalBroadcast(intent);
                break;
            case R.id.menu_add_to_playlist:
                Intent intentAdd = new Intent(mActivity, AddtoPlayListDialog.class);
                Bundle ardAdd = new Bundle();
                ardAdd.putSerializable("list", new ArrayList<>(Collections.singletonList(mSong.getId())));
                intentAdd.putExtras(ardAdd);
                mActivity.startActivity(intentAdd);
                break;
            case R.id.menu_add_to_play_queue:
                ToastUtil.show(mActivity, mActivity.getString(R.string.add_song_playqueue_success, MusicService.AddSongToPlayQueue(Collections.singletonList(mSong.getId()))));
                break;
            case R.id.menu_detail:
                mTag.detail();
                break;
            case R.id.menu_edit:
                mTag.edit();
                break;
            case R.id.menu_album_thumb:
                CustomThumb thumbBean = new CustomThumb(mSong.getAlbumId(), Constants.ALBUM, mSong.getAlbum());
                Intent thumbIntent = mActivity.getIntent();
                thumbIntent.putExtra("thumb", thumbBean);
                mActivity.setIntent(thumbIntent);
                Crop.pickImage(mActivity, Crop.REQUEST_PICK);
                break;
            case R.id.menu_ring:
                MediaStoreUtil.setRing(mActivity, mSong.getId());
                break;
            case R.id.menu_share:
                mActivity.startActivity(
                        Intent.createChooser(Util.createShareSongFileIntent(mSong, mActivity), null));
                break;
            case R.id.menu_delete:
                try {
                    String title = mActivity.getString(R.string.confirm_delete_from_playlist_or_library, mIsDeletePlayList ? mPlayListName : mActivity.getString(R.string.song_list));
                    Theme.getBaseDialog(mActivity)
                            .content(title)
                            .positiveText(R.string.confirm)
                            .negativeText(R.string.cancel)
                            .checkBoxPromptRes(R.string.delete_source, SPUtil.getValue(App.getContext(), SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.DELETE_SOURCE, false), null)
                            .onAny((dialog, which) -> {
                                if (which == POSITIVE) {
                                    boolean deleteSuccess = !mIsDeletePlayList ?
                                            MediaStoreUtil.delete(mSong.getId(), Constants.SONG, dialog.isPromptCheckBoxChecked()) > 0 :
                                            PlayListUtil.deleteSong(mSong.getId(), mPlayListName);

                                    ToastUtil.show(mActivity, deleteSuccess ? R.string.delete_success : R.string.delete_error);
                                }
                            })
                            .show();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
        }
        return true;
    }
}

File: app/src/main/java/tanzent/cassette/misc/menu/SongPopupListener.kt
package tanzent.cassette.misc.menu

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.PopupMenu
import android.view.MenuItem
import com.afollestad.materialdialogs.DialogAction.POSITIVE
import com.soundcloud.android.crop.Crop
import tanzent.cassette.App
import tanzent.cassette.R
import tanzent.cassette.bean.misc.CustomCover
import tanzent.cassette.bean.mp3.Song
import tanzent.cassette.db.room.DatabaseRepository
import tanzent.cassette.helper.DeleteHelper
import tanzent.cassette.request.network.RxUtil.applySingleScheduler
import tanzent.cassette.service.Command
import tanzent.cassette.service.MusicService.Companion.EXTRA_SONG
import tanzent.cassette.theme.Theme
import tanzent.cassette.ui.misc.Tag
import tanzent.cassette.ui.dialog.AddtoPlayListDialog
import tanzent.cassette.util.*
import tanzent.cassette.util.SPUtil.SETTING_KEY
import java.lang.ref.WeakReference
import kotlin.coroutines.experimental.coroutineContext

/**
 * Created by Remix on 2018/3/5.
 */

class SongPopupListener(activity: AppCompatActivity,
                        private val song: Song,
                        private val isDeletePlayList: Boolean,
                        private val playListName: String) : PopupMenu.OnMenuItemClickListener {
  private val tag = Tag(activity, song)
  private val ref = WeakReference(activity)

  override fun onMenuItemClick(item: MenuItem): Boolean {
    val activity = ref.get() ?: return true

    when (item.itemId) {
      R.id.menu_next -> {
        Util.sendLocalBroadcast(MusicUtil.makeCmdIntent(Command.ADD_TO_NEXT_SONG)
            .putExtra(EXTRA_SONG, song))
      }
      R.id.menu_add_to_playlist -> {
        AddtoPlayListDialog.newInstance(listOf(song.id))
            .show(activity.supportFragmentManager, AddtoPlayListDialog::class.java.simpleName)
      }
      R.id.menu_add_to_play_queue -> {
        DatabaseRepository.getInstance()
            .insertToPlayQueue(listOf(song.id))
            .compose(applySingleScheduler())
            .subscribe { it -> ToastUtil.show(activity, activity.getString(R.string.add_song_playqueue_success, it)) }
      }
      R.id.menu_detail -> {
        tag.detail()
      }
      R.id.menu_edit -> {
        tag.edit()
      }
      R.id.menu_album_thumb -> {
        val customCover = CustomCover(song.albumId, Constants.ALBUM,
            song.album)
        val coverIntent = activity.intent
        coverIntent.putExtra("thumb", customCover)
        activity.intent = coverIntent
        Crop.pickImage(activity, Crop.REQUEST_PICK)
      }
      R.id.menu_ring -> {
        MediaStoreUtil.setRing(activity, song.id)
      }
      R.id.menu_share -> {
        activity.startActivity(
            Intent.createChooser(Util.createShareSongFileIntent(song, activity), null))
      }
      R.id.menu_delete -> {
        val title = activity.getString(R.string.confirm_delete_from_playlist_or_library,
            if (isDeletePlayList) playListName else activity.getString(R.string.library))
        Theme.getBaseDialog(activity)
            .content(title)
            .positiveText(R.string.confirm)
            .negativeText(R.string.cancel)
            .checkBoxPromptRes(R.string.delete_source, SPUtil
                .getValue(App.getContext(), SETTING_KEY.NAME,
                    SETTING_KEY.DELETE_SOURCE, false), null)
            .onAny { dialog, which ->
              if (which == POSITIVE) {
                DeleteHelper
                    .deleteSong(song.id, dialog.isPromptCheckBoxChecked, isDeletePlayList, playListName)
                    .subscribe({ success -> ToastUtil.show(activity, if (success) R.string.delete_success else R.string.delete_error) }, { ToastUtil.show(activity, R.string.delete_error) })
                activity.startActivity(
                        Intent.createChooser(Util.refreshAdapterAndLibrary(activity), null)
                )
              }
            }
            .show()
      }
    }
    return true
  }
}
